### PR TITLE
i#3129 raw2trace perf: raw trace header inspection APIs

### PR DESCRIPTION
**Pull request by mtrofin**
Added thread start checking/validation APIs, and a related, general-purpose API for fetching out the timestamp out of a raw trace.

Issue: #3129

### CODE REVIEW

File: api/docs/release.dox [hunk 1]
```diff
@@ -245,6 +245,11 @@ Further non-compatibility-affecting changes include:
      and find_mapped_trace_address().
    - Added #trace_metadata_writer_t, a set of utility functions used by drcachesim's
      #raw2trace_t for writing trace metadata: process/thread ids, timestamps, etc.
+   - Added #trace_metadata_reader_t, a set of utilities for checking and validating
+     thread start successions of offline entries in a raw data file.
+ - Added drmemtrace_get_timestamp_from_offline_trace, an API for fetching the timestamp
```
Comment (derekbruening) [inline api/docs/release.dox]: I suggest checking the resulting docs: this is again missing links.  A function needs () to get a link.  Also up above it looks like the "replacing the following" functions are not being linked (in my build with doxygen 1.8.13): they likely need `raw2trace_t::` prefixes.

File: api/docs/release.dox [hunk 2]
```diff
@@ -245,6 +245,11 @@ Further non-compatibility-affecting changes include:
...
+   from the beginning of a raw trace bundle (regardless of whether it is a thread start
+   or just a subsequent bundle)
```
Comment (derekbruening) [inline api/docs/release.dox]: Missing `.`

File: clients/drcachesim/tests/raw2trace_io.cpp [hunk 3]
```diff
@@ -187,6 +187,62 @@ test_module_mapper(raw2trace_directory_t *dir)
...
 }
 
+bool
+test_trace_timestamp_reader(const raw2trace_directory_t *dir)
+{
+    std::istream *file = dir->thread_files[0];
+    // Seek back to the beginning to undo raw2trace_directory_t's validation
+    file->seekg(0);
+    offline_entry_t buffer[4];
+    file->read((char *)buffer, 4 * sizeof(offline_entry_t));
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_io.cpp]: BUFFER_SIZE_BYTES(buffer)

File: clients/drcachesim/tests/raw2trace_io.cpp [hunk 4]
```diff
@@ -187,6 +187,62 @@ test_module_mapper(raw2trace_directory_t *dir)
...
+
+    std::string error;
+    if (!trace_metadata_reader_t::is_thread_start(buffer, &error) && !error.empty())
+        return false;
+    uint64 timestamp = 0;
+    if (drmemtrace_get_timestamp_from_offline_trace(buffer, 4 * sizeof(offline_entry_t),
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_io.cpp]: BUFFER_SIZE_BYTES(buffer)

File: clients/drcachesim/tests/raw2trace_io.cpp [hunk 5]
```diff
@@ -187,6 +187,62 @@ test_module_mapper(raw2trace_directory_t *dir)
...
+    if (drmemtrace_get_timestamp_from_offline_trace(nullptr, 0, &timestamp2) ==
+        DRMEMTRACE_SUCCESS)
+        return false;
+    if (drmemtrace_get_timestamp_from_offline_trace(buffer, 0, &timestamp2) ==
+        DRMEMTRACE_SUCCESS)
+        return false;
+    if (drmemtrace_get_timestamp_from_offline_trace(buffer, sizeof(offline_entry_t),
+                                                    &timestamp2) == DRMEMTRACE_SUCCESS)
+        return false;
+    if (drmemtrace_get_timestamp_from_offline_trace(buffer, 4 * sizeof(offline_entry_t),
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_io.cpp]: BUFFER_SIZE_BYTES(buffer)

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 6]
```diff
@@ -1026,3 +1016,65 @@ raw2trace_t::~raw2trace_t()
...
+
+std::string
+trace_metadata_reader_t::check_entry_thread_start(const offline_entry_t *entry)
+{
+    std::string error;
+    if (is_thread_start(entry, &error))
+        return "";
+    if (error.empty())
+        return "Thread log file is corrupted: missing version entry";
+    ;
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: Looks like a mistake -- should be removed

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 7]
```diff
@@ -1026,3 +1016,65 @@ raw2trace_t::~raw2trace_t()
...
+    return error;
+}
+
+drmemtrace_status_t
+drmemtrace_get_timestamp_from_offline_trace(const void *trace, size_t size,
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: suggest: s/size/trace_size/

File: clients/drcachesim/tests/raw2trace_io.cpp [hunk 8]
```diff
@@ -230,6 +230,16 @@ test_trace_timestamp_reader(const raw2trace_directory_t *dir)
     if (timestamp != timestamp2)
         return false;
     REPORT("Verified boundary conditions");
+
+    offline_entry_t invalid_buffer[4];
+    memset(invalid_buffer, 0, 4 * sizeof(offline_entry_t));
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tests/raw2trace_io.cpp]: done

File: clients/drcachesim/tests/raw2trace_io.cpp [hunk 9]
```diff
@@ -230,6 +230,16 @@ test_trace_timestamp_reader(const raw2trace_directory_t *dir)
...
+    if (drmemtrace_get_timestamp_from_offline_trace(invalid_buffer,
+                                                    4 * sizeof(offline_entry_t),
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tests/raw2trace_io.cpp]: done

File: clients/drcachesim/tests/raw2trace_io.cpp (hunk 4 above)
Comment (mtrofin) [reply clients/drcachesim/tests/raw2trace_io.cpp]: done

File: clients/drcachesim/tests/raw2trace_io.cpp (hunk 5 above)
Comment (mtrofin) [reply clients/drcachesim/tests/raw2trace_io.cpp]: done

**Comment by mtrofin** [comment]
run arm tests
